### PR TITLE
Added a General library to orgs

### PR DESCRIPTION
**Pull request by ryanpbrewster**
It is a system library. It is created automatically when an org is created. Org
members are automatically added to the library when they join the org, and they
are removed when they leave or are kicked out.

The library cannot be modified or deleted. It automatically transfers ownership
when the org owner changes. It is deleted when the org is deleted.

@andrewconner @leogrim

### CODE REVIEW

**Comment by ryanpbrewster** [comment]
I want to take another look at this tomorrow once I've gotten some sleep. I feel like I had to make too many weird changes in order to get this to work.

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 1]
```diff
@@ -310,10 +310,11 @@ sealed abstract class LibraryKind(val value: String, val priority: Int) {
...
   case object SYSTEM_SECRET extends LibraryKind("system_secret", 1)
-  case object SYSTEM_PERSONA extends LibraryKind("system_persona", 2)
-  case object USER_CREATED extends LibraryKind("user_created", 2)
-  case object SYSTEM_READ_IT_LATER extends LibraryKind("system_read_it_later", 2)
-  case object SYSTEM_GUIDE extends LibraryKind("system_guide", 3)
+  case object SYSTEM_ORG_GENERAL extends LibraryKind("system_org_general", 2)
+  case object SYSTEM_PERSONA extends LibraryKind("system_persona", 3)
+  case object USER_CREATED extends LibraryKind("user_created", 3)
+  case object SYSTEM_READ_IT_LATER extends LibraryKind("system_read_it_later", 3)
+  case object SYSTEM_GUIDE extends LibraryKind("system_guide", 4)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala:317]: @leogrim I don't know why these have the numbers they do. It's a product question at heart.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationCommander.scala [hunk 2]
```diff
@@ -310,25 +311,30 @@ class OrganizationCommanderImpl @Inject() (
           val requester = userRepo.get(request.requesterId)
           organizationAnalytics.trackOrganizationEvent(org, requester, request)
 
-          val libsToReturn = libraryRepo.getBySpace(org.id.get, excludeState = None)
-          libsToReturn
+          val libsToDelete = libraryRepo.getBySpaceAndKind(org.id.get, LibraryKind.SYSTEM_ORG_GENERAL).map(_.id.get)
+          val libsToReturn = libraryRepo.getBySpaceAndKind(org.id.get, LibraryKind.USER_CREATED, excludeState = None).map(_.id.get)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationCommander.scala:315]: Maybe `libsToDelete = libRepo.getBySpace(org) -- libsToReturn`?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationMembershipCommander.scala [hunk 3]
```diff
@@ -212,35 +214,36 @@ class OrganizationMembershipCommanderImpl @Inject() (
...
-        Right(OrganizationMembershipAddResponse(request, newMembership))
+          savedMembership
+      }
     }
+
+    // Fire off a few Futures to take care of low priority tasks
+    maybeNotifySlackOfNewMember(request.orgId, request.targetId)
+    refreshOrganizationMembersTypeahead(request.orgId)
+
+    val orgGeneralLibrary = db.readOnlyReplica { implicit session => libraryRepo.getBySpaceAndKind(LibrarySpace.fromOrganizationId(request.orgId), LibraryKind.SYSTEM_ORG_GENERAL) }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationMembershipCommander.scala:238]: Do you think this should be `getBySpaceAndKind().head` or should I use a `foreach`? Until I backfill these libraries it's unsafe to use `.head`, but eventually should we move that way?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala [hunk 4]
```diff
@@ -425,21 +425,19 @@ class LibraryCommanderImpl @Inject() (
...
-      libraryMembershipRepo.getWithLibraryIdAndUserId(libId, lib.ownerId).foreach { oldOwnerMembership =>
-        libraryMembershipRepo.save(oldOwnerMembership.withState(LibraryMembershipStates.INACTIVE))
-      }
-      val existingMembershipOpt = libraryMembershipRepo.getWithLibraryIdAndUserId(libId, newOwner)
-      val newMembershipTemplate = LibraryMembership(libraryId = libId, userId = newOwner, access = LibraryAccess.OWNER)
-      libraryMembershipRepo.save(newMembershipTemplate.copy(id = existingMembershipOpt.map(_.id.get)))
-      libraryRepo.save(lib.withOwner(newOwner))
+    libraryMembershipRepo.getWithLibraryIdAndUserId(libraryId, lib.ownerId).foreach { oldOwnerMembership =>
+      libraryMembershipRepo.save(oldOwnerMembership.withState(LibraryMembershipStates.INACTIVE))
     }
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:436]: should the former owner be removed as a member here?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationMembershipCommander.scala (hunk 3 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationMembershipCommander.scala:238]: I'd leave it as is, if the general library doesn't exist it's probably not up to this function to figure it out, but to some integrity method that will also take care of recovering all existing organization members.

File: kifi-backend/modules/shoebox/app/com/keepit/model/FullLibraryInfo.scala [hunk 5]
```diff
@@ -81,6 +81,19 @@ case class LibraryCreateRequest(
...
+      name = "General",
+      visibility = LibraryVisibility.ORGANIZATION,
+      slug = "general",
+      kind = Some(LibraryKind.SYSTEM_ORG_GENERAL),
+      space = Some(LibrarySpace.fromOrganizationId(org.id.get)),
+      orgMemberAccess = Some(LibraryAccess.READ_WRITE)
+    )
+  }
+}
+
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/FullLibraryInfo.scala:96]: :+1:

File: kifi-backend/modules/shoebox/app/com/keepit/model/LibraryRepo.scala [hunk 6]
```diff
@@ -214,6 +215,20 @@ class LibraryRepoImpl @Inject() (
     }
   }
 
+  private def getByUserIdAndKind(userId: Id[User], kind: LibraryKind, excludeState: Option[State[Library]])(implicit session: RSession) = {
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryRepo.scala:218]: `getByOwnerIdAndKind` ?

**Comment by leogrim** [comment]
lgtm

**Comment by ryanpbrewster** [comment]
Sweet, thanks. I jumped the gun on merging a bit.

**Pull request merged**